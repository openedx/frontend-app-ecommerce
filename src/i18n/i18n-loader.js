/**
 * For each locale we want to support, react-intl needs 1) the locale-data, which includes
 * information about how to format numbers, handle plurals, etc., and 2) the translations, as an
 * object holding message id / translated string pairs.  A locale string and the messages object are
 * passed into the IntlProvider element that wraps your element hierarchy.
 *
 * Note that react-intl has no way of checking if the translations you give it actually have
 * anything to do with the locale you pass it; it will happily use whatever messages object you pass
 * in.  However, if the locale data for the locale you passed into the IntlProvider was not
 * correctly installed with addLocaleData, all of your translations will fall back to the default
 * (in our case English), *even if you gave IntlProvider the correct messages object for that
 * locale*.
 */

import { addLocaleData } from 'react-intl';
import Cookies from 'universal-cookie';

import arLocale from 'react-intl/locale-data/ar';
import enLocale from 'react-intl/locale-data/en';
import es419Locale from 'react-intl/locale-data/es';
import frLocale from 'react-intl/locale-data/fr';
import zhcnLocale from 'react-intl/locale-data/zh';

import COUNTRIES, { langs as countryLangs } from 'i18n-iso-countries';
import LANGUAGES, { langs as languageLangs } from '@cospired/i18n-iso-languages';

import arMessages from './messages/ar.json';
// no need to import en messages-- they are in the defaultMessage field
import es419Messages from './messages/es_419.json';
import frMessages from './messages/fr.json';
import zhcnMessages from './messages/zh_CN.json';

addLocaleData([...arLocale, ...enLocale, ...es419Locale, ...frLocale, ...zhcnLocale]);

// TODO: When we start dynamically loading translations only for the current locale, change this.
COUNTRIES.registerLocale(require('i18n-iso-countries/langs/ar.json'));
COUNTRIES.registerLocale(require('i18n-iso-countries/langs/en.json'));
COUNTRIES.registerLocale(require('i18n-iso-countries/langs/es.json'));
COUNTRIES.registerLocale(require('i18n-iso-countries/langs/fr.json'));
COUNTRIES.registerLocale(require('i18n-iso-countries/langs/zh.json'));

// TODO: When we start dynamically loading translations only for the current locale, change this.
// TODO: Also note that Arabic (ar) and Chinese (zh) are missing here.  That's because they're
// not implemented in this library.  If you read this and it's been a while, go check and see
// if that's changed!
LANGUAGES.registerLocale(require('@cospired/i18n-iso-languages/langs/en.json'));
LANGUAGES.registerLocale(require('@cospired/i18n-iso-languages/langs/es.json'));
LANGUAGES.registerLocale(require('@cospired/i18n-iso-languages/langs/fr.json'));

const messages = { // current fallback strategy is to use the first two letters of the locale code
  ar: arMessages,
  es: es419Messages,
  fr: frMessages,
  zh: zhcnMessages,
};

const cookies = new Cookies();

const getTwoLetterLanguageCode = code => code.substr(0, 2);

// Get the locale by setting priority. Skip if we don't support that language.
const getLocale = (localeStr) => {
  // 1. Explicit application request
  if (localeStr && messages[localeStr] !== undefined) {
    return localeStr;
  }
  // 2. User setting in cookie
  const cookieLangPref = cookies.get(process.env.LANGUAGE_PREFERENCE_COOKIE_NAME);
  if (cookieLangPref && messages[getTwoLetterLanguageCode(cookieLangPref)] !== undefined) {
    return getTwoLetterLanguageCode(cookieLangPref);
  }
  // 3. Browser language (default)
  return getTwoLetterLanguageCode(window.navigator.language);
};

const getMessages = (locale = getLocale()) => messages[locale];

const rtlLocales = ['ar', 'he', 'fa', 'ur'];
const isRtl = locale => rtlLocales.includes(locale);

const handleRtl = () => {
  if (isRtl(getLocale())) {
    document.getElementsByTagName('html')[0].setAttribute('dir', 'rtl');
    document.styleSheets[0].disabled = true;
  } else {
    document.styleSheets[1].disabled = true;
  }
};

/**
 * Provides a lookup table of country IDs to country names for the current locale.
 */
const getCountryMessages = (locale) => {
  const finalLocale = countryLangs().includes(locale) ? locale : 'en';

  return COUNTRIES.getNames(finalLocale);
};

/**
 * Provides a lookup table of language IDs to language names for the current locale.
 */
const getLanguageMessages = (locale) => {
  const finalLocale = languageLangs().includes(locale) ? locale : 'en';

  return LANGUAGES.getNames(finalLocale);
};

const sortFunction = (a, b) => {
  // If localeCompare exists, use that.  (Not supported in some older browsers)
  if (typeof String.prototype.localeCompare === 'function') {
    return a[1].localeCompare(b[1], getLocale());
  }
  if (a[1] === b[1]) {
    return 0;
  }
  // Otherwise make a best effort.
  return a[1] > b[1] ? 1 : -1;
};

/**
 * Provides a list of countries represented as objects of the following shape:
 *
 * {
 *   key, // The ID of the country
 *   name // The localized name of the country
 * }
 *
 * The list is sorted alphabetically in the current locale.
 * This is useful for select dropdowns primarily.
 */
const getCountryList = (locale) => {
  const countryMessages = getCountryMessages(locale);
  return Object.entries(countryMessages)
    .sort(sortFunction)
    .map(([code, name]) => ({ code, name }));
};

/**
 * Provides a list of languages represented as objects of the following shape:
 *
 * {
 *   key, // The ID of the language
 *   name // The localized name of the language
 * }
 *
 * The list is sorted alphabetically in the current locale.
 * This is useful for select dropdowns primarily.
 */
const getLanguageList = (locale) => {
  const languageMessages = getLanguageMessages(locale);
  return Object.entries(languageMessages)
    .sort(sortFunction)
    .map(([code, name]) => ({ code, name }));
};

export {
  getCountryList,
  getCountryMessages,
  getLanguageList,
  getLanguageMessages,
  getLocale,
  getMessages,
  handleRtl,
  isRtl,
};
