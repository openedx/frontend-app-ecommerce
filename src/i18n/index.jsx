import arMessages from './messages/ar.json';
import caMessages from './messages/ca.json';
// no need to import en messages-- they are in the defaultMessage field
import es419Messages from './messages/es_419.json';
import frMessages from './messages/fr.json';
import zhcnMessages from './messages/zh_CN.json';
import heMessages from './messages/he.json';
import idMessages from './messages/id.json';
import kokrMessages from './messages/ko_kr.json';
import plMessages from './messages/pl.json';
import ptbrMessages from './messages/pt_br.json';
import ruMessages from './messages/ru.json';
import thMessages from './messages/th.json';
import ukMessages from './messages/uk.json';

const messages = {
  ar: arMessages,
  'es-419': es419Messages,
  fr: frMessages,
  'zh-cn': zhcnMessages,
  ca: caMessages,
  he: heMessages,
  id: idMessages,
  'ko-kr': kokrMessages,
  pl: plMessages,
  'pt-br': ptbrMessages,
  ru: ruMessages,
  th: thMessages,
  uk: ukMessages,
};

export default messages;
