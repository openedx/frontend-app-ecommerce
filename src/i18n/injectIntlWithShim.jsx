import React from 'react';
import { injectIntl, intlShape } from 'react-intl';
import { logError } from '@edx/frontend-logging';


const injectIntlWithShim = (WrappedComponent) => {
  class ShimmedIntlComponent extends React.Component {
    static propTypes = {
      intl: intlShape.isRequired,
    };

    constructor(props) {
      super(props);
      this.shimmedIntl = Object.create(this.props.intl, {
        formatMessage: {
          value: (definition) => {
            if (definition === undefined || definition.id === undefined) {
              const error = new Error('i18n error: An undefined message was supplied to intl.formatMessage.');
              if (process.env.NODE_ENV !== 'production') {
                console.error(error); // eslint-disable-line no-console
                return '!!! Missing message supplied to intl.formatMessage !!!';
              }
              logError(error);
              return ''; // Fail silent in production
            }
            return this.props.intl.formatMessage(definition);
          },
        },
      });
    }

    render() {
      return <WrappedComponent {...this.props} intl={this.shimmedIntl} />;
    }
  }

  return injectIntl(ShimmedIntlComponent);
};


export default injectIntlWithShim;
