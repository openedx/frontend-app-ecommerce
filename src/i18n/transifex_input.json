{
  "siteheader.links.courses": "Courses",
  "siteheader.links.programs": "Programs & Degrees",
  "siteheader.links.schools": "Schools & Partners",
  "siteheader.user.menu.dashboard": "Dashboard",
  "siteheader.user.menu.profile": "Profile",
  "siteheader.user.menu.account.settings": "Account",
  "siteheader.user.menu.logout": "Logout",
  "siteheader.user.menu.login": "Login",
  "siteheader.user.menu.register": "Sign Up",
  "error.unexpected.message": "An unexpected error occurred.",
  "error.notfound.message": "The page you're looking for is unavailable or there's an error in the URL. Please check the URL and try again.",
  "root.page": "Congratulations!  You have a new micro-frontend.",
  "ecommerce.order.history.page.heading": "Order History",
  "ecommerce.order.history.no.orders": "Orders you place with edX will appear here.",
  "ecommerce.order.history.loading.orders": "Loading orders...",
  "ecommerce.order.history.loading.error": "Error: {error}"
}