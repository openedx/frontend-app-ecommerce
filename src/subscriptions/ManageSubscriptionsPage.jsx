import React, { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { FormattedMessage, useIntl } from '@edx/frontend-platform/i18n';
import { Button } from '@edx/paragon';

import { PageLoading, SupportLink } from '../components';

import { fetchStripeCustomerPortalURL } from './actions';
import { subscriptionsSelector } from './selectors';

const ManageSubscriptionsPage = () => {
  const { formatMessage } = useIntl();
  const dispatch = useDispatch();
  const { stripeCustomerPortalURL, stripeError } = useSelector(
    subscriptionsSelector,
  );

  const buttonLabel = formatMessage({
    id: 'ecommerce.order.history.manage.subscriptions.button',
    defaultMessage: 'Orders and subscriptions',
    description:
      'Button label to navigate to the orders and subscriptions page.',
  });

  useEffect(() => {
    dispatch(fetchStripeCustomerPortalURL());
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  useEffect(() => {
    if (stripeCustomerPortalURL) {
      window.location.href = stripeCustomerPortalURL;
    }
  }, [stripeCustomerPortalURL]);

  const renderLoading = () => (
    <PageLoading
      srMessage={formatMessage({
        id: 'ecommerce.order.history.loading.manage.subscriptions',
        defaultMessage: 'Loading manage subscriptions...',
        description: 'Message when loading the manage subscriptions page.',
      })}
    />
  );

  const renderError = () => (
    <div className="container-fluid d-flex py-5 justify-content-center align-items-start text-center">
      <div className="py-5 text-gray-700" style={{ maxWidth: '32em' }}>
        <FormattedMessage
          id="ecommerce.order.history.manage.subscriptions.error.heading"
          defaultMessage="Something went wrong"
          description="Heading when there is an error loading the manage subscriptions page."
        >
          {(text) => <h3 className="text-gray-700 mb-3">{text}</h3>}
        </FormattedMessage>
        <FormattedMessage
          id="ecommerce.order.history.manage.subscriptions.error.body"
          defaultMessage="The page youâ€™re looking for is unavailable or there was an error. Visit the {buttonLabel} page to manage your subscription."
          description="Message when there is an error loading the manage subscriptions page."
          values={{
            buttonLabel: <i className="text-lowercase">{buttonLabel}</i>,
          }}
        >
          {(text) => <p className="lead">{text}</p>}
        </FormattedMessage>
        <FormattedMessage
          tagName="p"
          id="ecommerce.order.history.manage.subscriptions.error.support"
          defaultMessage="If this problem persists, {supportLink}."
          description="Support Message when there is an error loading the manage subscriptions page."
          values={{
            supportLink: <SupportLink />,
          }}
        />
        <Button size="sm" href="/orders">
          {buttonLabel}
        </Button>
      </div>
    </div>
  );

  return stripeError ? renderError() : renderLoading();
};

export default ManageSubscriptionsPage;
