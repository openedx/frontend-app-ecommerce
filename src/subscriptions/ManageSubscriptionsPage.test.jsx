/* eslint-disable global-require */
import React from 'react';
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import { IntlProvider } from '@edx/frontend-platform/i18n';
import configureMockStore from 'redux-mock-store';

import ManageSubscriptionsPage from './ManageSubscriptionsPage';

const mockStore = configureMockStore();
const storeMocks = require('../store/__mocks__/mockStore');

describe('<ManageSubscriptions />', () => {
  describe('Renders correctly in various states', () => {
    it('renders when url is fetched correctly', () => {
      const tree = renderer
        .create(
          <IntlProvider locale="en">
            <Provider store={mockStore(storeMocks)}>
              <ManageSubscriptionsPage />
            </Provider>
          </IntlProvider>,
        )
        .toJSON();
      expect(tree).toMatchSnapshot();
    });
  });
});
