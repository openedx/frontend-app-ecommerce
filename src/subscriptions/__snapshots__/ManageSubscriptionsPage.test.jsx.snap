// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ManageSubscriptions /> Renders correctly in various states navigates when url is fetched correctly 1`] = `
<div>
  <div
    className="d-flex justify-content-center align-items-center flex-column"
    style={
      Object {
        "height": "50vh",
      }
    }
  >
    <div
      className="spinner-border text-primary"
      role="status"
    >
      <span
        className="sr-only"
      >
        Loading manage subscriptions...
      </span>
    </div>
  </div>
</div>
`;

exports[`<ManageSubscriptions /> Renders correctly in various states renders error ui when fetching url fails 1`] = `
<div
  className="container-fluid d-flex py-5 justify-content-center align-items-start text-center"
>
  <div
    className="py-5 text-gray-700"
    style={
      Object {
        "maxWidth": "32em",
      }
    }
  >
    <h3
      className="text-gray-700 mb-3"
    >
      Something went wrong
    </h3>
    <p
      className="lead"
    >
      The page youâ€™re looking for is unavailable or there was an error. Visit the 
      <i
        className="text-lowercase"
      >
        Orders and subscriptions
      </i>
       page to manage your subscription.
    </p>
    <p>
      If this problem persists, 
      <a
        className="pgn__hyperlink default-link standalone-link"
        href="http://localhost:18000/support/hc/requests/new"
        onClick={[Function]}
        target="_self"
      >
        contact support
      </a>
      .
    </p>
    <a
      className="btn btn-primary btn-sm"
      href="/orders"
      onClick={[Function]}
      onKeyDown={[Function]}
    >
      Orders and subscriptions
    </a>
  </div>
</div>
`;

exports[`<ManageSubscriptions /> Renders correctly in various states renders loading when url is being fetched 1`] = `
<div>
  <div
    className="d-flex justify-content-center align-items-center flex-column"
    style={
      Object {
        "height": "50vh",
      }
    }
  >
    <div
      className="spinner-border text-primary"
      role="status"
    >
      <span
        className="sr-only"
      >
        Loading manage subscriptions...
      </span>
    </div>
  </div>
</div>
`;
